File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/lettercombinations/LetterCombinations.java
package lettercombinations;

import java.util.*;

public class LetterCombinations {

    private static final String[] MAP = {
        "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
    };

    private final String digits;
    private final List<String> solutions = new ArrayList<>();

    public LetterCombinations(String digits) {
        if (digits == null || digits.isEmpty())
            throw new IllegalArgumentException("Digits must not be empty");
        this.digits = digits;
    }

    public void solveAndPrint() {
        backtrack(new StringBuilder(), 0);
        System.out.printf("Found %d combination(s) for digits \"%s\"%n%n", solutions.size(), digits);

        int limit = Math.min(solutions.size(), 8);
        for (int i = 0; i < limit; i++) {
            System.out.printf("#%d: %s%n", i + 1, solutions.get(i));
        }
        if (solutions.size() > limit) {
            System.out.printf("... and %d more.%n", solutions.size() - limit);
        }
    }

    private void backtrack(StringBuilder current, int index) {
        if (index == digits.length()) {
            solutions.add(current.toString());
            return;
        }
        String letters = MAP[digits.charAt(index) - '0'];
        for (char ch : letters.toCharArray()) {
            current.append(ch);
            backtrack(current, index + 1);
            current.deleteCharAt(current.length() - 1);
        }
    }
}

File: 29Items-Projects/7Algorithms/6-Backtracking/python/letter_combinations.py
"""Letter Combinations — generate all T9-style letter combinations for a phone number."""

_MAP = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]


class LetterCombinations:
    def __init__(self, digits: str):
        if not digits:
            raise ValueError("Digits must not be empty")
        self.digits = digits
        self.solutions: list[str] = []

    def solve_and_print(self) -> None:
        self._backtrack("", 0)
        print(f"Found {len(self.solutions)} combination(s) for digits \"{self.digits}\"\n")

        limit = min(len(self.solutions), 8)
        for i, sol in enumerate(self.solutions[:limit], 1):
            print(f"#{i}: {sol}")
        if len(self.solutions) > limit:
            print(f"... and {len(self.solutions) - limit} more.")

    def _backtrack(self, current: str, index: int) -> None:
        if index == len(self.digits):
            self.solutions.append(current)
            return
        for ch in _MAP[int(self.digits[index])]:
            self._backtrack(current + ch, index + 1)
